Multisrc preprocess

michnov opened the pull request:
### Support for multiple sources in the preprocessing

- multiple source languages allowed in `--source-lang` by argparse's `nargs="*"`
- if `--srcdict` is specified, the list of paths must be of the same length as `--source-lang` (again in the format that follows `nargs="*"`) with possibly repeating items
    - the final dictionary of several reapeating items is named after the language of the first such dictionary
    - e.g. if `--source-lang en1 en2 de` and `--srcdict d.en d.en d.de`, two final dictionaries are stored: `dict.en1.txt` and `dict.de.txt`
- if not, all languages that start with the same non-numeric prefix are treated as the same and a joint dictionary is thus made for them
    - the final dictionary is named using the common prefix of the languages
    - e.g. if `--source-lang en1 en2 de`, two final dictionaries are stored: `dict.en.txt` and `dict.de.txt`
- dealing with multiple sources together with alignment using the `--align-suffix` and `--alignfile` args is not supported yet

michnov committed:
Multi-source dictionaries handling

- multiple source languages delimited by "," allowed in --source-lang
- if --srcdict is specified, the list of paths must be of the same length as --source-lang (again delimited by ",") with possibly repeating items
    - the final dictionary of several reapeating items is named after the language of the first such dictionary
    - e.g. if --source-lang=en1,en2,de and --srcdict=d.en,d.en,d.de, two final dictionaries are stored: dict.en1.txt and dict.de.txt
- if not, all languages that start with the same non-numeric prefix are treated as the same and a joint dictionary is thus made for them
    - the final dictionary is named using the common prefix of the languages
    - e.g. if --source-lang=en1,en2,de, two final dictionaries are stored: dict.en.txt and dict.de.txt

michnov committed:
--source_langs and --srcdict values as nargs

- allow for space delimited list of multiple values automatically transformed to a list by argparse

michnov committed:
Support of multisrc in data binarization

varisd commented on fairseq_cli/preprocess.py:
```diff
@@ -45,6 +46,21 @@ def main(args):
+                raise RuntimeError(
+                    "Number of source languages ({}) must match the number of source dictionaries ({}).".format(
+                        len(args.source_lang), len(args.srcdict)))
+            args.srcdict_uniq = list(set(args.srcdict))
```
Cannot this mess up the order? (e.g., list(set([1,3,2,3,1])) outputs [1,2,3] - which is probably not desired)

michnov replied:
I don't think it messes up the order. I never overwrite `args.srcdict`. I store a set of unique dictionaries to `args.srcdict_uniq` (I admit I'd rather use a local variable instead of an `args` property) and create a `dict_index` as mapping from languages (in user-specified order) to the set of dictionaries. Then I load each dictionary only once on line 120 and map the dictionaries to languages using the `dict_index` on line 129.

varisd commented on fairseq_cli/preprocess.py:
```diff
@@ -45,6 +46,21 @@ def main(args):
+            return dict_index
+        else:
+            source_langs_stripped = [re.sub(r'\d+$', '', lang) for lang in args.source_lang]
+            args.srcdict_lang = list(set(source_langs_stripped))
```
Again, careful with reordering.

michnov replied:
Similar to the solution above. Instead of loading, using the set of unique languages, I build each dictionary only once on line 125 and map the dictionaries to original "languages" again on line 129.

varisd commented on fairseq_cli/preprocess.py:
```diff
@@ -72,14 +88,18 @@ def build_dictionary(filenames, src=False, tgt=False):
     if args.joined_dictionary:
         assert (
-            not args.srcdict or not args.tgtdict
+                not args.srcdict or not args.tgtdict
```
Is additional this indent necessary?

michnov replied:
This was unintentional, fixed. Thanks!

varisd reviewed (changes requested):
I am still not completely on board with the complicated naming of the srcdict and scr_lang args.
Technically, I would leave it to user to label the processed "languages" (it can be en1, en2 or en_a, en_b, or x, y).
The case where I have three source langs (a, b, c) and want to have a shared dict between "a" and "b" can be solved outside of fairseq by first calling preprocess.py on "a+b" corpus and "c" corpus, creating "a+b" and "c" dictionary and then calling the preprocess again with src_dicts "a => a+b", "b => a+b" and "c". Therefore, I do not think that this feature does not need to be explicitly supported.

michnov committed:
Refactored to follow Dusan's suggestions

- if --srcdict is not specified, a specific dictionary is created for every source language (i.e. every item in the --source-lang list), e.g. if --source-lang x y z, three final dictionaries are stored: dict.{x,y,z}.txt
- if --srcdict is specified, the list of paths must be of the same length as --source-lang with possibly repeating items
     * the output dictionary of several reapeating items is named as a concatenation of the corresponding languages delimited by "@", e.g. if --source-lang x y z and --srcdict d.xz d.y d.xz, two final dictionaries are stored: dict.x@z.txt and dict.y.txt

varisd commented on fairseq_cli/preprocess.py:
```diff
@@ -337,12 +360,12 @@ def make_all_alignments():
+                    "alignment.{}-{}.txt".format(args.source_lang, args.target_lang),
+                ),
+                "w",
+                encoding="utf-8",
```
no need for extra indent

varisd commented on fairseq_cli/preprocess.py:
```diff
@@ -141,7 +163,7 @@ def merge_result(worker_result):
             pool = Pool(processes=num_workers - 1)
             for worker_id, (start_offset, end_offset) in enumerate(
-                more_chunks, start=1
+                    more_chunks, start=1
```
no need for extra indent

varisd commented on fairseq_cli/preprocess.py:
```diff
@@ -208,7 +230,7 @@ def merge_result(worker_result):
             pool = Pool(processes=num_workers - 1)
             for worker_id, (start_offset, end_offset) in enumerate(
-                more_chunks, start=1
+                    more_chunks, start=1
```
no need for extra indent

varisd commented on fairseq_cli/preprocess.py:
```diff
@@ -115,7 +131,13 @@ def build_dictionary(filenames, src=False, tgt=False):
-    src_dict.save(dict_path(args.source_lang))
+    for i, d in enumerate(src_dict_uniq):
+        if args.srcdict:
+            # lang = args.source_lang[dict_index.index(i)]
```
remove comment

varisd commented on fairseq_cli/preprocess.py:
```diff
@@ -115,7 +131,13 @@ def build_dictionary(filenames, src=False, tgt=False):
+    for i, d in enumerate(src_dict_uniq):
+        if args.srcdict:
+            # lang = args.source_lang[dict_index.index(i)]
+            lang = [args.source_lang[li] for li, di in enumerate(dict_index) if di == i]
```
shouldn't lang be scalar?

varisd reviewed (changes requested):
Ready for merge after slight modifications (see the comments bellow).
Please, pay extra attention to the comment regarding the ``lang;; variable assignment.
